Further Fluid Tank Implementation

34will opened the pull request:
# Description

Added the following functionality to Fluid Tanks:

- The ability to fill/drain from the Block in the world, using a Bucket or something that implements/has the capability of `IFluidHandlerItem`
- Fluid Tanks now contain the `IFluidHandlerItem` capability
- Fixed rendering of the Fluid Tank item
- Added animation & sound to the filling & draining of the Fluid Tank
- Made it so that empty (and default configured) Fluid Tank Blocks when broken now stack with other empty (and default configured) Fluid Tank Items.

Possible Breaking Changes:

- Changes to the `MachineBlockEntity` mean the IOConfig, Redstone Control, and Inventory data is only written to NBT if it needs to be (i.e. is non-default), so if something expects it to always be there, it now may not be
- Changes to the `copyNBT` method of `MachinesLootTable` to use the new `ImprovedContextNbtProvider` may affect other things, and may not be futureproof. The `ImprovedContextNbtProvider` cannot simply extend the `ContextNbtProvider` class because it `private`s a lot of its underlying functionality (I implemented `ImprovedContextNbtProvider` by just duplicating the `ContextNbtProvider` class and editing the parts I needed to) so if changes occur to `ContextNbtProvider` in the future, they may not be reflected in this code. I don't know if this is something we could submit a PR to Forge for, to expose the class more?

# Checklist:

- [x] My code follows the style guidelines of this project (.editorconfig, most IDEs will use this for you)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

<!-- Thanks to: https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ for the building blocks of this template -->

34will committed:
Modified FluidTankBlockEntity to expose some methods for filling and draining it.

34will committed:
Added FluidTankBlock to provide a block for the Fluid Tank so that it can be interacted with in the world.
Modified MachineBlocks to register the above for the Fluid Tank and Pressurized Fluid Tank.

34will committed:
Added FluidHandlerBlockItemStack to provide a FluidHandlerItemStack that correctly handles BlockItems, namely reading and writing from the BlockEntityTag.
Modified FluidTankItem to provide the above as a capability, to provide different capacities for the standard Fluid Tank vs the Pressurized, and to display the fluid contained in the tooltip.
Modified MachineBlocks to register the relevant FluidTankItem subclasses for the Fluid Tank and Pressurized Fluid Tank.

34will committed:
Modified FluidTankBlockEntity and FluidTankBEWLR to correctly load (and write) the NBT data

34will committed:
Added Animation to provide a class for handling a simple Animation.
Modified FluidTankBER to animate the fluid when added or removed.
Modified FluidTankBlockEntity to register and deregister itself after being loaded and broken, respectively.
Modified EIOMachines to faciliatate the above.

34will committed:
Formatting

34will committed:
Modified FluidTankBlock to play a sound when filling or emptying a bucket.

34will committed:
Modified FluidTankBlock to play sounds better, and trigger events.

34will committed:
Modified FluidTankItem to run initCapabilities like BucketItem.

34will committed:
Modified FluidTankBlock to check if the itemStack is empty before doing anything.

34will committed:
Styling

34will committed:
Added NbtTags to provide a centralized location for NBT Tags.
Modified PaintedBlockColor, PaintedSandEntity, PaintedSlabBlockItem, PaintUtils, FallingMachineEntity, MachinesLootTable, FluidHandlerBlockItemStack, FluidTankBEWLR to reflect the above.

34will committed:
Modified IIOConfig to include an isDefault method, to check if the current config is different from the default.
Modified FixedIOConfig and IOConfig to implement the above.

34will committed:
Modified MachineBlockEntity to only write the IOConfig, Redstone Control, and Inventory to NBT if required.

34will committed:
Added MachineNbtTags to provide a centralized location for NBT Tags used in the Machine mod.
Modified MachineBlockEntity to use the above.

34will committed:
Formatting.

34will committed:
Added ImprovedContextNbtProvider to provide an NbtProvider that saves BlockEntities without the id and positional metadata.
Added EIOLootNbtProvider to register the above.
Modified EnderIO to facilitate the above.

34will committed:
Modified MachinesLootTable to use the new ImprovedContextNbtProvider.

34will committed:
Adding generated files from runData.

34will committed:
Merge commit '425d1a64ac0d2cc19f4766268e0b64db715880ea' into feature/fluid_tank

34will committed:
Adding generated files from runData.

Rover656 commented on src/api/java/com/enderio/api/io/IIOConfig.java:
```diff
@@ -56,4 +56,9 @@ default void cycleMode(Direction side) {
+    /**
+     * Whether the IO config is unchanged from initial values.
+     */
+    boolean isDefault();
```
Personally not a fan of having this method, but I'm unsure of a way that would alleviate the need for it

Rover656 commented on src/decor/java/com/enderio/decoration/client/render/PaintedBlockColor.java:
```diff
@@ -39,8 +40,8 @@ public int getColor(BlockState state, @Nullable BlockAndTintGetter level, @Nulla
     public int getColor(ItemStack itemStack, int tintIndex) {
-        if (itemStack.getTag() != null && itemStack.getTag().contains("BlockEntityTag")) {
-            CompoundTag blockEntityTag = itemStack.getTag().getCompound("BlockEntityTag");
+        if (itemStack.getTag() != null && itemStack.getTag().contains(NbtTags.BLOCK_ENTITY_NBT_KEY)) {
```
Might be worth investigating if there's a vanilla constant for this, but I do like the idea of keeping all our tags together

justliliandev replied:
There is only a private one in BlockItem

Rover656 commented on src/machines/java/com/enderio/machines/EIOMachines.java:
```diff
@@ -32,6 +34,8 @@ public static void onConstruct(FMLConstructModEvent event) {
         MachineRecipes.register();
         MachineCapacitorKeys.register();
+
+        MinecraftForge.EVENT_BUS.register(FluidTankBlockEntity.class);
```
Can you use the annotations instead of registering manually here please

34will replied:
Fixed in 7de7cbaaa8a9c24b9bea48518feedc2b6fafac79 and 11535800ec28e4b58dcd4a4123647bda116bc039

Rover656 commented on src/machines/java/com/enderio/machines/client/rendering/blockentity/FluidTankBER.java:
```diff
@@ -12,61 +15,136 @@
     @Override
-    public void render(FluidTankBlockEntity blockEntity, float partialTick, PoseStack poseStack, MultiBufferSource bufferSource, int packedLight,
-        int packedOverlay) {
+    public void render(FluidTankBlockEntity blockEntity, float partialTick, PoseStack poseStack,
```
Could I ask for more documentation around the animation implementation you've added?

34will replied:
Yes of course, do you mean just comments and Javadocs, or actual documentation in like a wiki or something?

Rover656 replied:
Yeah just comments and javadocs please

34will replied:
Fixed in 3010b205c2c2779c487f29d5f68ad5f32ac068e6

Rover656 commented on src/machines/java/com/enderio/machines/client/rendering/blockentity/FluidTankBER.java:
```diff
@@ -12,61 +15,136 @@
+                fluidHeight, color);
+    }
+
+    public static void addBlock(FluidTankBlockEntity entity) {
```
I'm not a massive fan of tracking things like this, but again I understand why its here

Rover656 commented on src/machines/java/com/enderio/machines/common/blockentity/FluidTankBlockEntity.java:
```diff
@@ -90,66 +110,144 @@ public void serverTick() {
         super.serverTick();
     }
 
+    private IFluidHandlerItem GetIFluidHandlerItem(ItemStack itemStack) {
```
Maybe move this into a static utility class like the EnergyUtil?

justliliandev replied:
also fix naming. Methods start with a lowercase words

34will replied:
Fixed in ccb0d52f862fb553321db570a897ecd4b6c7b7db

Rover656 commented on src/machines/java/com/enderio/machines/common/blockentity/FluidTankBlockEntity.java:
```diff
@@ -90,66 +110,144 @@ public void serverTick() {
+        return drainTankWithBucket(null);
+    }
+
+    public Pair<Boolean, FluidStack> drainTankWithBucket(Function<FluidStack, Boolean> shouldDrain) {
```
Mark shouldDrain as nullable if you're allowing null

34will replied:
Fixed in 06c076d25b74df9093933ecb397774aed2d55f4f

Rover656 commented on src/machines/java/com/enderio/machines/common/blockentity/FluidTankBlockEntity.java:
```diff
@@ -162,30 +260,37 @@ public FluidTank getFluidTank() {
         super.saveAdditional(pTag);
-        pTag.put("fluid", fluidTank.writeToNBT(new CompoundTag()));
+        if (fluidTank.isEmpty()) {
+            pTag.remove(FluidHandlerBlockItemStack.FLUID_NBT_KEY);
```
is it necessary to try calling remove? As far as my understanding goes, whenever save is called its a fresh tag, so existing keys shouldn't be present?

Rover656 commented on src/machines/java/com/enderio/machines/common/blockentity/base/MachineBlockEntity.java:
```diff
@@ -411,28 +417,45 @@ public void saveAdditional(CompoundTag pTag) {
 
         if (inventory != null) {
-            pTag.put("inventory", inventory.serializeNBT());
+            boolean hasContents = false;
```
I don't think its necessary to do this check, for the very little storage it saves it'll add a bunch of time onto the saving process

34will replied:
The reason for this check is that when the `BlockEntity` is converted over into an `Item` after being broken, it retains the extra information in the NBT data. So, for example, when a Fluid Tank is spawned/crafted into the inventory, its item NBT data is `{id: "enderio:fluid_tank", Count: 1b}`. When it is placed into the world, then immediately broken again, its item NBT data (without my changes) becomes: `{id: "enderio:fluid_tank", Count: 1b, tag: {BlockEntityTag: {x: ..., y: ..., z: ..., io_config: ..., redstone: ..., inventory: ..., id: "enderio:fluid_tank"}}}`. This has the annoying (to me at least), although admittedly not game-breaking, issue of the fact that that Fluid Tank item will not stack with another empty Fluid Tank, despite no logical reason not to.

This was simply the first way I came across of (partly) solving the issue. So I could move the functionality across into the NBT copying? But would probably need a specific provider for the Machine blocks, in the event that other things need to use that functionality.

I'll have to defer to your better judgement about the about the time in the saving process, but I could try to simplify the checks to just simple boolean checks by updating whether each item is empty/default when they are changed? Also isn't it better to potentially not save & transmit data, even at the slight increased complexity/time cost?

Rover656 commented on src/machines/java/com/enderio/machines/common/item/FluidTankItem.java:
```diff
@@ -26,4 +58,32 @@ public BlockEntityWithoutLevelRenderer getCustomRenderer() {
+    @Nullable
+    @Override
+    public ICapabilityProvider initCapabilities(ItemStack stack, @Nullable CompoundTag nbt) {
+        if (this instanceof FluidTankItem) {
```
Why this check? We know this is a FluidTankItem

34will replied:
I just copied this from the `BucketIem` where they do the same check. It does seem odd, so happy to remove.

Rover656 commented on src/main/java/com/enderio/base/common/loot/providers/nbt/ImprovedContextNbtProvider.java:
```diff
@@ -0,0 +1,140 @@
+ * {@linkplain LootContextParams#BLOCK_ENTITY block entity}'s NBT data or an
+ * entity's NBT data based on an {@link LootContext.EntityTarget} but without the id and positional data.
+ */
+public class ImprovedContextNbtProvider implements NbtProvider {
```
What is the motive behind needing this over the old system?

34will replied:
As with the other checks above the reason for this is that when the `BlockEntity` is converted over into an Item after being broken, it retains extra unneeded `BlockEntity` information in the NBT data. So, for example, when a Fluid Tank is spawned/crafted into the inventory, its item NBT data is `{id: "enderio:fluid_tank", Count: 1b}`. When it is placed into the world, then immediately broken again, its item NBT data (without this class) becomes: `{id: "enderio:fluid_tank", Count: 1b, tag: {BlockEntityTag: {x: ..., y: ..., z: ..., id: "enderio:fluid_tank"}}}`. Again, this has the annoying issue of the Fluid Tank item not stacking with another empty Fluid Tank.

This issue comes from the `ContextNbtProvider` calling `saveWithFullMetadata` on the `BlockEntity`, writing the `x`, `y`, `z`, and `id` to the NBT Tag. This provider is a carbon copy of the `ContextNbtProvider` class, except it calls `saveWithoutMetadata` on the `BlockEntity`, and returns `null` if that returns `null`. I would have just extended the `ContextNbtProvider` class, but the relevant methods to override are `private`.

As I mentioned above, the other exclusion of tags being written could be performed here as well.

Rover656 commented on src/core/java/com/enderio/core/client/rendering/Animation.java:
```diff
@@ -0,0 +1,44 @@
+
+import net.minecraft.util.Mth;
+
+public class Animation {
```
Javadocs please :)

34will replied:
Fixed in cb65ecffc869a1b00025ff0c587219e4f69e6153

Rover656 reviewed (changes requested):
Just a quick skim through, a number of comments and changes. I'm not a fan of how some of the animation stuff is working in terms of tracking all the block entities in a map; might be worth looking into other ways of doing this (maybe storing the animation state on the BE instead somehow).

34will committed:
Modified EIOMachines to remove the FluidTankBlockEntity registering.
Modified FluidTankBlockEntity to self-register the events via Annotations.

34will committed:
Fixed wrong Bus.

34will committed:
Added FluidUtil to provide a helper class for dealing with Fluids.
Modified FluidTankBlockEntity to use the above.

34will committed:
Modified FluidTankBlockEntity to include @Nullable in all the relevant methods.

34will committed:
Modified Animation to include JavaDocs.

34will committed:
Modifed FluidTankBER to include JavaDocs and comments

34will commented:
> I'm not a fan of how some of the animation stuff is working in terms of tracking all the block entities in a map; might be worth looking into other ways of doing this (maybe storing the animation state on the BE instead somehow).

I'm not sure why I didn't store the animation information in the BE in the first place 😅 I will give that a go.

34will committed:
Modified Animation.isComplete to function using the timer rather than the value.

34will committed:
Modified FluidTankBlockEntity to store and initialise an AnimationInformation.
Modified FluidTankBER to remove the map used to store information about animations, and instead use the AnimationInformation stored inside the FluidTankBlockEntity.
Modified AnimationInformation to use getters & setters.

34will committed:
Modified FluidTankItem.initCapabilities to remove the odd check.

34will committed:
Formatting

34will committed:
Modified FluidUtil to include methods for draining and filling FluidTanks.
Modified FluidTankBlockEntity and MachineBlockEntity to use the above.
Modified FluidTankBlock to use the above.

34will committed:
Modified FluidTankBlockEntity.saveAdditional to remove the unneccessary check.

justliliandev commented on src/core/java/com/enderio/core/client/rendering/Animation.java:
```diff
@@ -0,0 +1,85 @@
+// TODO: Support more interpolation methods, if required.
+/**
+ * Class for animating between two values over an amount of time, using
+ * {@link Mth.lerp}.
```
Fix link using # and put in one line
```suggestion
 * Class for animating between two values over an amount of time, using {@link Mth#lerp}.
```

justliliandev commented on src/core/java/com/enderio/core/client/rendering/Animation.java:
```diff
@@ -0,0 +1,85 @@
+    private float target;
+    private float current;
+    private float currentDurationTick;
+    private float totalDurationTick;
```
make start, target and totalDurationTick final

justliliandev commented on src/machines/java/com/enderio/machines/client/rendering/blockentity/FluidTankBER.java:
```diff
@@ -16,57 +19,156 @@
+    public static class AnimationInformation {
+        private int currentTarget;
+        private Animation animation = null;
+        private Fluid fluid = null;
```
fluid will be assigned in the constructor, don't do it here

justliliandev commented on src/machines/java/com/enderio/machines/client/rendering/blockentity/FluidTankBER.java:
```diff
@@ -16,57 +19,156 @@
+            this.currentTarget = currentTarget;
+        }
+
+        public Animation getAnimation() {
```
mark as nullable

justliliandev commented on src/machines/java/com/enderio/machines/client/rendering/blockentity/FluidTankBER.java:
```diff
@@ -16,57 +19,156 @@
+            return fluid;
+        }
+
+        public void setFluid(@Nullable Fluid fluid) {
```
fluids shouldn't be nullable but rather Fluids.EMPTY

justliliandev commented on src/machines/java/com/enderio/machines/client/rendering/blockentity/FluidTankBER.java:
```diff
@@ -16,57 +19,156 @@
+        private Animation animation = null;
+        private Fluid fluid = null;
+
+        public AnimationInformation(@Nullable Fluid fluid, int currentTarget) {
```
not Nullable

justliliandev commented on src/machines/java/com/enderio/machines/client/rendering/blockentity/FluidTankBER.java:
```diff
@@ -16,57 +19,156 @@
+
+    public static class AnimationInformation {
+        private int currentTarget;
+        private Animation animation = null;
```
mark nullable

justliliandev commented on src/machines/java/com/enderio/machines/client/rendering/blockentity/FluidTankBER.java:
```diff
@@ -16,57 +19,156 @@
+
+                information.setCurrentTarget(fluidAmount);
+                frameTarget = information.getAnimation().getCurrent();
+            } catch (Exception e) {
+                // Catching exception from Animation constructor, just render as normal.
+                frameTarget = fluidAmount / capacity;
+            }
```
Don't use exceptions for controlflow, the exception will never be thrown here, as ANIMATION_TICKS is positive

34will replied:
So would you prefer that I don't throw the Exception in the `Animation` constructor? Is there a way for it to bypass the Exception with an Annotation or something given that we know `ANIMATION_TICKS` is greater than zero? Sorry, I'm not versed in the deeper nuances of Java

justliliandev replied:
throw an illegalargumentexception, don't declare that the constructor throws something and remove the catch

justliliandev commented on src/machines/java/com/enderio/machines/common/block/FluidTankBlock.java:
```diff
@@ -0,0 +1,202 @@
+                sound = optionalSound.get();
+            } else {
+                sound = SoundEvents.BUCKET_FILL;
+            }
```
```suggestion
            sound = fluid.getPickupSound().orElse(SoundEvents.BUCKET_FILL);
```

justliliandev commented on src/machines/java/com/enderio/machines/common/block/FluidTankBlock.java:
```diff
@@ -0,0 +1,202 @@
+
+        Optional<IFluidHandlerItem> fluidHandlerCap = itemStack
+                .getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY).resolve();
+        if (!fluidHandlerCap.isPresent()) {
```
use fluidHandlerCap.isEmpty

justliliandev commented on src/machines/java/com/enderio/machines/common/block/FluidTankBlock.java:
```diff
@@ -0,0 +1,202 @@
+        // Read the information for the Fluid in the item.
+        boolean itemIsEmpty = false;
+        for (int i = 0; i < fluidHandler.getTanks(); i++) {
+            itemIsEmpty = fluidHandler.getFluidInTank(i).getFluid() == Fluids.EMPTY;
```
```suggestion
            itemIsEmpty = fluidHandler.getFluidInTank(i).isEmpty();
```

justliliandev commented on src/machines/java/com/enderio/machines/common/block/FluidTankBlock.java:
```diff
@@ -0,0 +1,202 @@
+        Fluid fluid;
+        // If the block is empty, and the item is not empty: fill the block from the
+        // item.
+        if (blockIsEmpty && !itemIsEmpty) {
```
can be simplified to blockIsEmpty, because if blockIsEmpty, item can't be empty, because of the return ealier

justliliandev commented on src/machines/java/com/enderio/machines/common/block/FluidTankBlock.java:
```diff
@@ -0,0 +1,202 @@
+            fluid = fluidTankBlockEntity.getFluidTank().getFluid().getFluid();
+            // If the block is not empty, and the item is empty: fill the item from the
+            // block.
+        } else if (!blockIsEmpty && itemIsEmpty) {
```
can be simplified to itemIsEmpty, because block can't be empty, because of previous returns

justliliandev commented on src/machines/java/com/enderio/machines/common/block/FluidTankBlock.java:
```diff
@@ -0,0 +1,202 @@
+                    } else {
+                        playPlaceSound(pPlayer, pLevel, result.getMiddle(), pPos);
+                    }
+                case FAIL:
```
the fallthorugh sems to handle this correctly, but is hard to read, add `return result.getLeft();` to the first case

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/FluidTankBlockEntity.java:
```diff
@@ -48,8 +51,8 @@ public Enhanced(BlockEntityType<?> type, BlockPos worldPosition, BlockState bloc
 
     private final FluidTank fluidTank;
-
     private final MachineFluidHandler fluidHandler;
```
move fluidHandler to a local variable in the constructor, as it's only used there

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/FluidTankBlockEntity.java:
```diff
@@ -48,8 +51,8 @@ public Enhanced(BlockEntityType<?> type, BlockPos worldPosition, BlockState bloc
     private final FluidTank fluidTank;
-
     private final MachineFluidHandler fluidHandler;
+    private AnimationInformation animationInformation;
```
add Nullable annotation and add UseOnly(LogicalSide.CLIENT)

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/FluidTankBlockEntity.java:
```diff
@@ -63,7 +66,16 @@ public FluidTankBlockEntity(BlockEntityType<?> type, BlockPos worldPosition, Blo
+        addDataSlot(new FluidStackDataSlot(fluidTank::getFluid, this::setFluidStack, SyncMode.WORLD));
+    }
+
+    private void setFluidStack(FluidStack stack) {
```
Add UseOnly(LogicalSide.CLIENT) Annotation

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/FluidTankBlockEntity.java:
```diff
@@ -159,33 +192,45 @@ public FluidTank getFluidTank() {
         return fluidTank;
     }
 
+    public AnimationInformation getAnimationInformation() {
```
Add Nullable Annotation

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/base/MachineBlockEntity.java:
```diff
@@ -238,7 +242,7 @@ public boolean canAct() {
     public boolean canActSlow() {
         return canAct()
-            && level.getGameTime() % 5 == 0;
+                && level.getGameTime() % 5 == 0;
```
undo change

34will replied:
Apologies, this is because I am coding with VSCode, and the `.editorconfig` values don't seem to be being respected. I did download IntelliJ to check, and it put this on one line, which is different from what this was originally. I have just left it as one line.

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/base/MachineBlockEntity.java:
```diff
@@ -211,7 +214,8 @@ protected void onContentsChanged(int slot) {
      * @apiNote Must call this on custom MachineInventory handlers!
      */
-    protected void onInventoryContentsChanged(int slot) { }
+    protected void onInventoryContentsChanged(int slot) {
+    }
```
undo change

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/base/MachineBlockEntity.java:
```diff
@@ -251,7 +255,8 @@ public boolean canActSlow() {
     private void forceResources() {
         for (Direction direction : Direction.values()) {
             if (ioConfig.getMode(direction).canForce()) {
-                // TODO: Maybe some kind of resource distributor so that items are transmitted evenly around? rather than taking the order of Direction.values()
+                // TODO: Maybe some kind of resource distributor so that items are transmitted
+                // evenly around? rather than taking the order of Direction.values()
```
undo change

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/base/MachineBlockEntity.java:
```diff
@@ -394,8 +385,10 @@ protected void clearCaches() {
+            itemHandlerCache.put(direction, addInvalidationListener(
+                    neighbor.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, direction.getOpposite())));
+            fluidHandlerCache.put(direction, addInvalidationListener(
+                    neighbor.getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, direction.getOpposite())));
```
undo change

34will replied:
Again, this is because I am coding with VSCode, my apologies. I did format it with IntelliJ which gave something different to both, so I have left that in.

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/base/MachineBlockEntity.java:
```diff
@@ -411,28 +404,45 @@ public void saveAdditional(CompoundTag pTag) {
-            pTag.put("inventory", inventory.serializeNBT());
+            boolean hasContents = false;
+            int stacks = inventory.getSlots();
+            for (int i = 0; i < stacks; i++) {
```
add this to MachineInventory
```
public boolean isEmpty() {
    return stacks.stream().noneMatch(ItemStack::isEmpty);
}
```
so this can be replaced with !inventory.isEmpty()

justliliandev commented on src/machines/java/com/enderio/machines/common/blockentity/base/MachineBlockEntity.java:
```diff
@@ -456,7 +466,8 @@ public Component getDisplayName() {
         if (this.level.getBlockEntity(this.worldPosition) != this)
             return false;
-        return pPlayer.distanceToSqr(this.worldPosition.getX() + 0.5D, this.worldPosition.getY() + 0.5D, this.worldPosition.getZ() + 0.5D) <= 64.0D;
+        return pPlayer.distanceToSqr(this.worldPosition.getX() + 0.5D, this.worldPosition.getY() + 0.5D,
+                this.worldPosition.getZ() + 0.5D) <= 64.0D;
```
undo change

justliliandev commented on src/main/java/com/enderio/base/common/loot/providers/nbt/ImprovedContextNbtProvider.java:
```diff
@@ -0,0 +1,141 @@
+        }
+
+        public Set<LootContextParam<?>> getReferencedContextParams() {
+            return ImmutableSet.of(LootContextParams.BLOCK_ENTITY);
```
use Set.of

34will replied:
The `ContextNbtProvider` that I shamelessly copied this from uses this, so I'm reluctant to change it, but if you think it's okay, then I will.

justliliandev commented on src/main/java/com/enderio/base/common/loot/providers/nbt/ImprovedContextNbtProvider.java:
```diff
@@ -0,0 +1,141 @@
+    private static ImprovedContextNbtProvider.Getter forEntity(final LootContext.EntityTarget pEntityTarget) {
+        return new ImprovedContextNbtProvider.Getter() {
+            @Nullable
+            public Tag get(LootContext p_165589_) {
```
name parameter

justliliandev commented on src/main/java/com/enderio/base/common/loot/providers/nbt/ImprovedContextNbtProvider.java:
```diff
@@ -0,0 +1,141 @@
+            }
+
+            public Set<LootContextParam<?>> getReferencedContextParams() {
+                return ImmutableSet.of(pEntityTarget.getParam());
```
use Set.of

34will replied:
The `ContextNbtProvider` that I shamelessly copied this from uses this, so I'm reluctant to change it, but if you think it's okay, then I will.

justliliandev replied:
it's effectivly the same, and at that point it's personal preference. ImmutableSet could indicate, that someone wants implementationdetails for a specific reason, while the set just indicates, that you want a set

justliliandev commented on src/main/java/com/enderio/base/common/loot/providers/nbt/ImprovedContextNbtProvider.java:
```diff
@@ -0,0 +1,141 @@
+        Set<LootContextParam<?>> getReferencedContextParams();
+    }
+
+    public static class InlineSerializer implements GsonAdapterFactory.InlineSerializer<ImprovedContextNbtProvider> {
+        public JsonElement serialize(ImprovedContextNbtProvider p_165597_, JsonSerializationContext p_165598_) {
+            return new JsonPrimitive(p_165597_.getter.getId());
+        }
+
+        public ImprovedContextNbtProvider deserialize(JsonElement p_165603_, JsonDeserializationContext p_165604_) {
+            String s = p_165603_.getAsString();
+            return ImprovedContextNbtProvider.createFromContext(s);
+        }
+    }
```
This seems to be unused, if there is likely a future reason for it, keep it

justliliandev commented on src/main/java/com/enderio/base/common/loot/providers/nbt/ImprovedContextNbtProvider.java:
```diff
@@ -0,0 +1,141 @@
+        }
+    }
+
+    public static class Serializer
+            implements net.minecraft.world.level.storage.loot.Serializer<ImprovedContextNbtProvider> {
```
put on one line, rename Serializer to JsonSeriailzer or something, so you can import it

justliliandev commented on src/main/java/com/enderio/base/common/loot/providers/nbt/ImprovedContextNbtProvider.java:
```diff
@@ -0,0 +1,141 @@
+        /**
+         * Serialize the value by putting its data into the JsonObject.
+         */
+        public void serialize(JsonObject p_165610_, ImprovedContextNbtProvider p_165611_,
+                JsonSerializationContext p_165612_) {
+            p_165610_.addProperty("target", p_165611_.getter.getId());
+        }
+
+        /**
+         * Deserialize a value by reading it from the JsonObject.
+         */
+        public ImprovedContextNbtProvider deserialize(JsonObject p_165618_, JsonDeserializationContext p_165619_) {
```
name parameter

34will committed:
Formatting and finals for Animation

34will committed:
Nullables for FluidTankBER

34will committed:
Tidy ups in FluidTankBlock

34will committed:
Tidy up in FluidTankBlock

34will committed:
Nullable and UseOnly for FluidTankBlockEntity.

34will committed:
MachineFluidHandler change in FluidTankBlockEntity.

34will committed:
Undone formatting for MachineBlockEntity.

34will committed:
Modified MachineInventory to include an isEmpty method.
Modified MachineBlockEntity.saveAdditional to use the above.

34will committed:
Tidy up ImprovedContextNbtProvider.
Modified EIOLootNbtProvider to reflect the above.

34will committed:
Merge branch 'dev/1.19.x' into feature/fluid_tank

34will committed:
Missed close bracket

34will committed:
Modified FluidTank Items to  use new ForgeCapabilities.

34will committed:
Merge error

34will committed:
Merge commit '9914000971cf004fb0b965e0c9201990d5408044' into feature/fluid_tank

34will committed:
Fixed Animation to just throw an IllegalArgumentException, and not declare it.
Modified FluidTankBER to reflect the above.

34will committed:
Modified ImprovedContextNbtProvider to use Set instead of ImmutableSet.
